test/tla/Paxos.tla

```diff
@@ -7,63 +7,49 @@
 (* This is a specification of the Paxos algorithm without explicit leaders *)
 (* or learners.  It refines the spec in Voting                             *)
 (***************************************************************************)
-EXTENDS Integers
+EXTENDS Integers, Variants
 -----------------------------------------------------------------------------
-(***************************************************************************)
-(* The constant parameters and the set Ballots are the same as in Voting.  *)
-(***************************************************************************)
-
 (*
-CONSTANT Value, Acceptor, Quorum
+ @typeAlias: MESSAGE =
+        M1a({ bal: Int })
+      | M1b({ acc: A, bal: Int, mbal: Int, mval: V })
+      | M2a({ bal: Int, val: V })
+      | M2b({ acc: A, bal: Int, val: V })
+    ;
+ *)
+Paxos_typedefs == TRUE
 
-ASSUME QuorumAssumption == /\ \A Q \in Quorum : Q \subseteq Acceptor
-                           /\ \A Q1, Q2 \in Quorum : Q1 \cap Q2 # {}
+\* we accompany the MESSAGE type with a constructor for each case
 
-Ballot ==  Nat
+\* @type: Int => MESSAGE;
+M1a(bal) ==
```

Comment:
Do type annotations for variants have to be generic? Wouldn't this more precisely be
```
 \* @type: Int => M1a({ bal: Int });
```

Reply:
Yes, you have to generalize the type to `MESSAGE`. Otherwise, the variant is locked to only one option (no row variables) and you cannot compose it with another variant.